src/components/who-is-online/index.ts

```diff
@@ -0,0 +1,240 @@
+
+  /**
+   * @function subscribeToRealtimeEvents
+   * @description Unsubscribes to realtime events
+   * @returns {void}
+   */
+  private unsubscribeToRealtimeEvents(): void {
+    this.realtime.presenceMouseParticipantJoinedObserver.unsubscribe(this.onParticipantJoined);
+    this.realtime.presenceMouseParticipantLeaveObserver.unsubscribe(this.onParticipantLeave);
+    this.realtime.participantsObserver.unsubscribe(this.onParticipantListUpdate);
+  }
+
+  private compareParticipants() {
+    const realtimeParticipants = this.realtime.getParticipants;
+
+    for (let i = 0; i < this.participants.length; i++) {
+      if (!realtimeParticipants[this.participants[i].id]) return false;
+    }
+
+    return true;
+  }
+
+  /**
+   * @function subscribeToRealtimeEvents
+   * @description Receives data about participants in the room who were not loaded when the component was initialized
+   * @param {Record<string, AblyParticipant>} data
+   * @returns {void}
+   */
+  private onParticipantListUpdate = (data: Record<string, AblyParticipant>) => {
+    const updatedParticipants = Object.values(data);
+    if (updatedParticipants.length === this.participants.length && this.compareParticipants())
+      return;
+
+    const currentParticipants: string[] = this.participants.map((participant) => participant.id);
+
+    this.participants = [
+      ...this.participants,
+      ...updatedParticipants
+        .filter((participant: AblyParticipant) => {
+          return !currentParticipants.includes(participant.data.id);
+        })
+        .map((participant: AblyParticipant) => {
+          const { slotIndex } = participant.data;
+          const color = MeetingColorsHex[slotIndex];
+          return { ...participant.data, color };
+        }),
+    ] as Participant[];
+
+    this.element.participants = this.participants;
+  };
+
+  /**
+   * @function onParticipantJoined
+   * @description Updates the participants list when a new participant joins the meeting
+   * @param {Ably.Types.PresenceMessage} participant
+   * @returns {void}
+   */
+  private onParticipantJoined = (data: Participant) => {
+    const participant = this.realtime.getParticipants[data.id]?.data;
+
+    const alreadyInList = participant
+      ? this.participants?.find((element) => {
+          return element.id === participant.id;
+        })
+      : false;
+
+    const color = MeetingColorsHex[participant?.slotIndex];
+    data.slotIndex = participant?.slotIndex;
+
+    if (alreadyInList) {
+      this.participants = this.participants.map((participant) => {
+        if (participant.id === data.id) {
+          return {
+            ...participant,
+            ...data,
+            color,
+          };
+        }
+        return participant;
+      });
+    } else {
+      this.participants.push({ ...this.realtime.getParticipants[data.id]?.data, ...data, color });
+    }
+    if (!color) this.getColorAfterDelay(data.id);
+
+    this.element.participants = this.participants;
+  };
```

Comment:
Is better to use some instead of find and skip if participant is not in realtime list


```suggestion
  /**
   * @function onParticipantJoined
   * @description Updates the participants list when a new participant joins the meeting
   * @param {Ably.Types.PresenceMessage} participant
   * @returns {void}
   */
  private onParticipantJoined = (data: Participant) => {
    const participant = this.realtime.getParticipants[data.id]?.data;

   if(!participant) return;
         
    const alreadyInList =  this.participants?.some((element) => {
          return element.id === participant.id;
    });

    const color = MeetingColorsHex[participant?.slotIndex];
    data.slotIndex = participant?.slotIndex;

    if (alreadyInList) {
      this.participants = this.participants.map((participant) => {
        if (participant.id === data.id) {
          return {
            ...participant,
            ...data,
            color,
          };
        }
        return participant;
      });
    } else {
      this.participants.push({ ...this.realtime.getParticipants[data.id]?.data, ...data, color });
    }
    if (!color) this.getColorAfterDelay(data.id);

    this.element.participants = this.participants;
  };
```

Reply:
If I'm not mistaken, it's important not to return early because the call to getColorAfterDelay deals with some rare cases in which the recently joined participant appears to not have a color for themselves.
![image](https://github.com/SuperViz/sdk/assets/111044527/2f12ca81-65da-4ca5-bee2-f2f2991033f0)

Reply:
Yes, you can ignore the skip early, but the other points like using some instead of find should be considered.